Upgrade to Bootstrap 4

myfrom opened the pull request:
Did all the necessary changes to upgrade the web app to Bootstrap 4. As far as I tested, everything should work fine.

Fixes #200

sushain97 commented:
It looks like the build is failing? Also, this diff is massive! I think that the themes didn't previously require as much CSS to be present in this repo. I think it was downloaded during the build process. Is it possible for us to revert to that setup?

sushain97 commented:
I think the themes are required (not the full CSS) since they contained app-specific color overrides. Can you take a look at the old versions?

sushain97 commented on assets/css/general.css:
```diff
@@ -38,19 +38,33 @@ h2 {
+/* Utilities */
+
+/* sass-lint:disable class-name-format */
+.position-relative {
```
Does BS4 not already come with an equivalent? I was under the impression that a lot of these utilities could be removed.

myfrom replied:
I removed the unnecessary ones in latest push

sushain97 commented:
No need to squash commits. I'll do that on merge.

myfrom commented:
Oh, I understand the themes now. So they are overwrites if we build specifically for a theme. I could revert them, the problem might be that I don't think that 3.x themes are compatible with Bootstrap 4 and the new 4.x themes have some new ones and some old ones are missing.

sushain97 commented:
The overrides are just for app-specific stuff though. They're not dependent on bs3 and if they are, it's only a couple lines to change?

myfrom commented:
Yeah, but the set of themes for v3 isn't identical to the set for v4. E.g. theme called 'paper' no longer exists, I think it got replaced by 'materia' and I'm not sure if they have the same look. Also, there are a few new themes, which weren't there before. I'm not sure what to do with those.

I'll definitely revert the themes folder and do a bit more research. If there were just renames, I'll try to rename the overwrites where applicable and I think it would be good to open another issue to add overwrites for new themes where needed. How about that?

sushain97 commented:
Ohhh. Don't worry about old themes. For new themes, just add similar overrides. I'm pretty sure the overrides just copy over some colors.

myfrom commented:
Ok, thanks! I'll do it tomorrow when I'm back from school.

Could you also have a look at the site, please? It should be good (excluding the themes) but there were subtle changes, mostly related to how Bootstrap handles spacing now. I don't think they're breaking changes and actually they might even be better for usability on mobile phones so I left them with defaults. It's a few pixels of a difference compared to current master.

sushain97 commented:
Took a look at the translator interface and it's pretty close! There are however a couple of mostly minor issues. I'll take a look at the other interfaces after this one is fixed :)

https://601-28576409-gh.circle-artifacts.com/0/home/circleci/project/build/index.eng.html?dir=arg-cat#translation

- random extra space after "Help us improve..."

![image](https://user-images.githubusercontent.com/4211838/48741845-4c6c2880-ec22-11e8-89d8-69562258a7d4.png)

- install notice always appears (some JS/CSS class mismatch)

![image](https://user-images.githubusercontent.com/4211838/48741838-47a77480-ec22-11e8-85df-4bef6742b4ce.png)

- hovering over long language names cuts them off 

![image](https://user-images.githubusercontent.com/4211838/48741835-42e2c080-ec22-11e8-9bc3-f4bb04a60e6c.png)

- the disabled language gray is _way_ too close to normal black

![image](https://user-images.githubusercontent.com/4211838/48741878-66a60680-ec22-11e8-87c0-e22b1b2a0045.png)

- clicking on a source language results in a weird orange bar flickering

- large vertical spacing between options (not a huge deal)

![image](https://user-images.githubusercontent.com/4211838/48741853-555cfa00-ec22-11e8-8377-517f5a7e3f72.png)

myfrom commented:
The issues you mentioned, including themes should be fixed now.

myfrom commented:
Thank you for the additional time!

I have fixed the first two issues and also adjusted the vertical spacing on the checkboxes.
When it comes to the nav buttons, they could have the hover effect either by adding additional (non-Bootstrap) styles, tell me if I should add that.

But could you also guide me how to enable those beta changes on the local build? I can't get the navbar entries that you were talking about, even when building from master. Am I missing some special make command?

JPJPJPOPOP commented:
@myfrom You enable the navbar and more modes through the config.conf file.

myfrom commented:
Fixed.

The styling of nav element has changed visibly in Bootstrap 4 - especially if you look at the top navigation (from beta) - but as per your previous comment, I left it unchanged.

sushain97 commented:
Some minor concerns before I think this is good for task approval. I'm not necessarily going to merge it yet because there are some more minor tweaks that I think it'll need but that can be another task (feel free to take it up).

1. Some overflow issues in destination language dropdown (does this happen in current version too?)

![image](https://user-images.githubusercontent.com/4211838/48888760-9272e900-ede8-11e8-8b81-71e59cc4a4eb.png)

2. I can click on a disabled language in the destination language dropdown and it turns orange. Is it not properly disabled to prevent click events?

3. It should be possible to fix the float here so that the navbar is directly below the locale selector (see beta.apertium.org). I think this is just an artifact of the new styles that requires fixing rather than a new look.

![image](https://user-images.githubusercontent.com/4211838/48888901-f9909d80-ede8-11e8-9f55-1ae8b61f798e.png)

sushain97 commented:
Oh no! An RTL locale (like Hebrew) is super messed up. Let me know if this would be really complicated to fix due to some dependency issues. Otherwise, it needs to be remedied.

![image](https://user-images.githubusercontent.com/4211838/48888988-3a88b200-ede9-11e8-9d4a-9274b97f12e7.png)

sushain97 commented:
If it's pretty involved, we can leave it for another task/issue since I think you've done a fair bit of work on this already.

myfrom commented:
> Some overflow issues in destination language dropdown (does this happen in current version too?)

I haven't been able to test with this set of languages but if I put in there random text it overflows on live version too. I added text ellipsis in my edits.

> I can click on a disabled language in the destination language dropdown and it turns orange. Is it not properly disabled to prevent click events?

I removed Bootstrap's `dropdown-item` from those items. Seems to work fine now without it and the class was causing a few style issues.

> It should be possible to fix the float here

I changed it up a bit so it should be nicer now. I don't want to use floats here because of a few things related to layout that Bootstrap 4 changed so I manipulate bottom padding of the navbar depending on the size of the device. Should behave similarly to original but I can't guarantee that on every locale.
I can try to make it behave like Bootstrap 3 if you want.

On the RTL version, it should be fixable with a few additional styles. I think it shouldn't take too long. I'll start working on that.

sushain97 commented:
> it overflows on live version too. I added text ellipsis in my edits.

It seems to wrap here which is better IMO:

![image](https://user-images.githubusercontent.com/4211838/48917978-874aa800-ee3e-11e8-955d-6b750d7a4cc7.png)

> I removed Bootstrap's dropdown-item from those items. Seems to work fine now without it and the class was causing a few style issues.

Sounds good!

> I can try to make it behave like Bootstrap 3 if you want.

I wouldn't spend more than 15 minutes on it overall. Like I said, I just want the broad strokes to be resolved here.

> On the RTL version, it should be fixable with a few additional styles. I think it shouldn't take too long. I'll start working on that.

Awesome!

sushain97 commented:
Ah, looks good!

myfrom commented:
RTL should be fine now. There's a bug with language selection dropdowns but it's probably related to JavaScript, not the styles and it also appears on live version.

sushain97 commented on html/index.html.in:
```diff
@@ -58,7 +60,7 @@
         <!-- [if lt IE 10]>
             <style type="text/css">
-                button#translateDoc { display: none; }
+                button#translateDoc { display: none;; }
```
```suggestion
                button#translateDoc { display: none; }
```

sushain97 commented:
I'm going to accept this task since I think you've done a great job with it. There are things that need to be cleaned up for a merge of this as v4 of the app. If you're interested, let me know and I can create a new task.

e.g. the alignment here

![image](https://user-images.githubusercontent.com/4211838/48962831-ef7eb400-ef3a-11e8-9525-039829dd1890.png)

There are some other inconsistencies such as some extra stuff being enabled in mobile that doesn't fully work on mobile.

The one thing I do request is for the random double semicolon and a description of how you generated this bootstrap.css file so that it can be duplicated or something similar in the future.

myfrom commented:
Thank you!

I've used [this tool](https://bootstrap.build/app) to build the theme. I didn't think of exporting it as `_variables.scss` tho 😬 I'll reproduce it and share the file so it can be loaded onto the program to make updates easier in the future.

I was talking about a bug that makes all the languages go into one column in RTL locale:
![www apertium org_](https://user-images.githubusercontent.com/19410489/48971866-06e39d00-f020-11e8-8b76-c7be09b1e274.png)
That's a screenshot from current live version. PS: I've just noticed it appears only on smaller screen width.

sushain97 commented:
> I've used this tool to build the theme. I didn't think of exporting it as _variables.scss tho 😬 I'll reproduce it and share the file so it can be loaded onto the program to make updates easier in the future.

Awesome, thanks!

> That's a screenshot from current live version. PS: I've just noticed it appears only on smaller screen width.

Huh, nice find. Do you mind creating an issue?

myfrom commented:
Here's the file:
[apertium-variables-scss.zip](https://github.com/apertium/apertium-html-tools/files/2613421/apertium-variables-scss.zip)
I packed it into a ZIP so it can be uploaded into GitHub.

And I opened an issue for that bug - it's #333

sushain97 commented:
@myfrom - awesome! lmk if you want me to make a task that involves cleaning up the rest of the details here.

MaathavanJkr commented:
So what bit works are not finished?

sushain97 commented:
I honestly don't remember. You'd have to try it. If you don't see anything that doesn't match up with the current site, I can take a look.

MaathavanJkr commented:
> I honestly don't remember. You'd have to try it. If you don't see anything that doesn't match up with the current site, I can take a look.

I cant find any changes except size differents. May be im not familiar with it. Can you please take a look at it?

MaathavanJkr commented:
@sushain97 Can you please. :)

sushain97 commented:
There are actually a bunch of minor things mentioned in this thread.

MaathavanJkr commented:
@sushain97 After a analysis in this pull, These are my output but still i cant find which isnt fixed yet. :(

> random extra space after "Help us improve..."

fixed

> install notice always appears (some JS/CSS class mismatch)

Fixed.

> hovering over long language names cuts them off
the disabled language gray is way too close to normal black

FIxed.

> clicking on a source language results in a weird orange bar flickering

Fixed.

> large vertical spacing between options (not a huge deal)

Fixed.

> I'm afraid we really need to decrease the vertical spacing here or maybe the size (see current version)

Fixed.

> This should be flipping to the left? Maybe some JS is mismatched?

Fixed.

> Some overflow issues in destination language dropdown (does this happen in current version too?)

Fixed.

> I can click on a disabled language in the destination language dropdown and it turns orange. Is it not properly disabled to prevent click events?

Fixed.

> It should be possible to fix the float here so that the navbar is directly below the locale selector (see beta.apertium.org). I think this is just an artifact of the new styles that requires fixing rather than a new look.

Fixed.

> Oh no! An RTL locale (like Hebrew) is super messed up. Let me know if this would be really complicated to fix due to some dependency issues. Otherwise, it needs to be remedied.

Fixed.

> Would be nice if we could get this hover background back but it's not 100% necessary if Bootstrap doesn't support it anymore.

Its talked not needed.
